---
title: "Rosalind"
author: "Erik Sundberg"
date: "6/6/2019"
output: html_document
---

```{r packages, echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(tidyverse)
library(edgeR)
library(seqinr)
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


####**Reverse complementary sequence**
```{r reverse complement, echo=TRUE, message=TRUE}

s <- aseq#"AAAACCCGGT"

rc <- function(x) {
#Split a character string
ss <- unlist(strsplit(x, NULL))

#Reverse the split character string
ssrev <- rev(ss)

#Replace characters with complementary letters
for(i in 1:length(ssrev)) {
  if(ssrev[i] == "T") { 
    ssrev[i] <- "A" 
  } else if(ssrev[i] == "A") {
      ssrev[i] <- "T"
  } else if(ssrev[i] == "G") {
      ssrev[i] <- "C"
  } else if(ssrev[i] == "C") {
      ssrev[i] <- "G"
    }
}
#Paste the complementary string
sc <- paste(ssrev, collapse = "")
sc }
```


####**Fibonacci sequence**
```{r Rabbits and recurrence}
rF <- c()
n = 32
k = 3
i = 0
for(i in 1:n) {
  if(i == 1) {
    rF[i] <- 1
  } else if(i == 2) {
    rF[i] <- 1
  } else rF[i] = rF[i-1] + k*rF[i-2]
}
rF[length(rF)]



```


####**Calculate percentage GC-content in a DNA sequence**
```{r GC-content}

#Function for calculating GC-content from text in fasta format
#x is the filename, eg 'rosalind_gc.txt'
gcc <- function(x) {
  rs <- read_file(x)
  rs2 <- unlist(strsplit(rs, ">"))
  rs2 <- rs2[-1]
  rs3 <- gsub("\n", "", rs2)
  i = 0
  v <- c()
  nv <- c()
  for(i in 1:length(rs3)) {
    s1 <- unlist(strsplit(rs3[i], NULL))
    sn <- paste(s1[1:13], collapse = "")
    s1 <- s1[14:length(s1)]
    s1 <- as.factor(s1)
    s1s <- summary(s1)
    gc1 <- unname(s1s["G"]) + unname(s1s["C"])
    gc1c <- gc1/length(s1) * 100 
    v[i] <- gc1c
    nv[i] <- sn
  }
  names(v) <- nv
  print(noquote(names(v[which(v == max(v))])))
  return(max(v))
}


#Alternatively, using seqinr package
fasta <- read.fasta('rosalind_gc.txt')
gc_content <- apply(matrix(names(fasta)), 1, function(x){GC(fasta[[x]])})
most_gc <- which(gc_content==max(gc_content))

#Result
rbind(names(fasta)[most_gc], paste(signif(gc_content[most_gc], 7) * 100, "%", sep=""))
```


####Counting points mutations
```{r Hamming distance}

a <- c("")
a2 <- unlist(strsplit(a, "\n"))

b1 <- unlist(strsplit(a2[1], ""))
b2 <- unlist(strsplit(a2[2], ""))
length(which(b1 != b2))

```


####Mendelian inheritance
```{r Mendelian inheritance}

vec <- c("19 18 28")
vec <- as.double(unlist(strsplit(vec, " ")))

#Total nr of organisms
norg = sum(vec)

#Creat a matrix for probabilities
ofal <- c("AA","Aa","aa")
mmat <- matrix(nrow = 3, ncol = 3)
rownames(mmat) <- ofal
colnames(mmat) <- ofal
#for loop for the probability of mating combinations
for(i in 1:nrow(mmat)){
  for(j in 1:ncol(mmat)) {
    if(vec[i] == vec[j]) {
      mmat[i,j] = vec[i]/norg*(vec[j]-1)/(norg-1)
      } else if(vec[i] != vec[j]) {
      mmat[i,j] = (vec[i]/(norg))*(vec[j]/(norg-1))
    }
  }
}

dPr = sum(mmat[1,1:3])+sum(mmat[2:3,1])+0.75*sum(mmat[2,2])+sum(mmat[2,3])
dPr

#Alternative solutions:
gt.counts <- c(19,18,28)
gt.probs <- gt.counts / sum(gt.counts);
gt.subcounts <- t(matrix(gt.counts,3,3)) - diag(3);
gt.subprobs <- gt.subcounts / rowSums(gt.subcounts);

dominant.mat <- matrix(c(1,1,1,1,0.75,0.5,1,0.5,0),3,3);
sum(rowSums(gt.subprobs * dominant.mat) * gt.probs);

#or
domi <- rep(NA,1000000)
pop <- c(rep("AA",16), rep("Aa",17),rep("aa",15))
for(x in 1:length(domi)) {
  pair <- sample(pop,2)
  domi[x] <- sample(strsplit(pair[1],NULL)[[1]],1) == "A" | sample(strsplit(pair[2],NULL)[[1]],1) == "A"
}
sum(domi)/length(domi)


```

####Translating RNA into Protein
```{r Translating RNA into Protein}
#Given file w RNA string
s <- readLines("/Users/eriksundberg/Downloads/rosalind_prot.txt")
s2 <- gsub("\n", "", s)
#RNA codon table
RC <- read_table('RNAcodonTable.txt', col_names = FALSE)
RCvec <- unlist(c(RC[,c(1,3,5,7)]))
RCvec <- unname(RCvec)
names(RCvec) <- unname(unlist(c(RC[,c(2,4,6,8)])))

ss <- c()
i = 1
j = 1
while(i < nchar(s) && j <= nchar(s)/3) {
  ss[j] <- substr(s, i,i+2)
  i <- i+3
  j <- j+1
}
#Amino-acid vector
pvec <- c()
k=1
for(k in 1:length(ss)) {
  if(ss[k] %in% RCvec) {
    pvec[k] <- names(which(RCvec == ss[k]))
  }
}
pvec <- pvec[-k]
paste(pvec, collapse = "")


#Alternative solution:
## The translation table
codon_trans <- scan( what=character(), textConnection("
UUU F      CUU L      AUU I      GUU V
UUC F      CUC L      AUC I      GUC V
UUA L      CUA L      AUA I      GUA V
UUG L      CUG L      AUG M      GUG V
UCU S      CCU P      ACU T      GCU A
UCC S      CCC P      ACC T      GCC A
UCA S      CCA P      ACA T      GCA A
UCG S      CCG P      ACG T      GCG A
UAU Y      CAU H      AAU N      GAU D
UAC Y      CAC H      AAC N      GAC D
UAA Stop   CAA Q      AAA K      GAA E
UAG Stop   CAG Q      AAG K      GAG E
UGU C      CGU R      AGU S      GGU G
UGC C      CGC R      AGC S      GGC G
UGA Stop   CGA R      AGA R      GGA G
UGG W      CGG R      AGG R      GGG G 
"
))

lc <- length( codon_trans )
codon_trans_table <- codon_trans[ seq(1, lc, by=2) ]
names( codon_trans_table ) <- codon_trans[ seq(2, lc, by=2) ]

## Handy replace function
kReplace <- function( vec, orig, out=names(orig) ) {
  tmp <- out[ match( vec, orig ) ]
  tmp[ is.na(tmp) ] <- vec[ is.na(tmp) ]
  tmp
}

## Slice function
strslice <- function(x, n) {
  x <- as.data.frame( stringsAsFactors=FALSE, 
                      matrix( unlist( strsplit( x, "" ) ), ncol=n, byrow=T )
  )

  do.call( paste0, x )

}

dat <- scan( "rosalind_prot.txt", what=character(), quiet=TRUE )

codons <- strslice(dat, 3)
cat(paste(collapse="",
      kReplace( codons, codon_trans_table )
))

```

####Finding Motif's in DNA
```{r Finding Motifs in DNA}
s <- readLines("/Users/eriksundberg/Downloads/rosalind_subs.txt")
s1 <- s[1]
s2 <- s[2]
unlist(gregexpr(pattern = s2, text = s1, perl=T))

s2 <- unlist(strsplit("CTTCGTTTCCTGCTCGTTTCTCGTTTCTCAATCGTTTCATCGTGTCGTTTCTGTCGTTTCAGGTCGTTTCTCGTTTCTTCGTTTCTCGTTTCGTTATCGTTTCAATCGTTTCGTCTCGTTTCCTTCCCCAAGGTTCGTTTCTATCGTTTCCCTCGTTTCTGTCGTTTCGATATTTCGTTTCTCGTTTCACCACATCGTTTCATTCGTTTCATCGTTTCTCGTTTCTTCGTTTCTCGTTTCTCGTTTCTCGTTTCCGTCGTTTCTCGTTTCGTGATTCGTTTCTCGTTTCCTAATATCGTTTCACCATCGTCGTTTCATCGTCTTTCGTTTCACTCGTTTCGTCTGTTCGTTTCTCGTTTCTCGTTTCAGTCGTTTCCGTCGTTTCCTATCGTTTCGTCGTTTCTCGTTTCGCAAAGACATCGTTTCCATGTCGTTTCTCGTTTCTCGTTTCACATCGTTTCTCGTTTCCGTCGTTTCTCTCGTTTCCTATCGTTTCCCGTCGTTTCCCCTCGTTTCAACTCGTTTCATCGTTTCTCGTTTCGTCGTTTCAGTCGTTTCTCGTTTCCGGAATCGTTTCTCGTTTCCTCGTTTCGGTATCGTTTCTCTCGTTTCCCATCGTTTCTCGTTTCCTAGCTCGTTTCTCGTTTCGCAATCCCTACACTCGTTTCGTTCGTTTCATCGTTTCATTCGTTTCGTCGTTTCCTCGTTTCGATCGTTTCTAGATCGTTTCTCGTTTCTCGTTTCGGCAAGTCGTTTCTCGTTTCTTGACAAGTCGTTTCATCGTTTCTTTGCTTTTCGTTTCTCATTCGTTTCTCGTTTCTCGCATTCGTTTCGCTTCGTTTCCTCGTTTCCCCGGGTCTTTCGTTTCACTTAACTCGTTTCTCGTTTCCTCGTTTCTCGTTTCCTCGTTTC", ""))
t <- unlist(strsplit("TCGTTTCTC", ""))
i = 1
ix = c()
nix = c()
for( i in 1:length(s2) ) {
  end <- length(t) - 1 + i
  if(paste(t, collapse = "") == paste(s2[i:end], collapse = "")) {
    ix <- c(ix, i)
  } else nix <- c(nix, i)
}

ix

```

####Consensus and Profile (most likely common ancestor)
```{r Consensus and Profile}
DNAstr <- readLines("/Users/eriksundberg/Downloads/rosalind_cons.txt")
DNApst <- paste(DNAstr, collapse = "")
DNAsub <- gsub(">Rosalind_[0-9][0-9][0-9][0-9]", " ", DNApst)
DNAspl <- unlist(strsplit(DNAsub, " "))
DNAspl2 <- DNAspl[2:11]
DNAmat <- sapply(DNAspl2, strsplit, "", USE.NAMES = FALSE)
DNAmatr <- unname(t(as.data.frame(DNAmat[1:10])))

Prof <- matrix(nrow = 4, ncol = ncol(DNAmatr))
rownames(Prof) <- levels(factor(DNAmatr))

for(i in 1:ncol(DNAmatr)) {
  Prof[,i] <- summary(factor(DNAmatr[,i], levels = c("A","C","G","T")))
}

Cons <- c()
for(i in 1:ncol(DNAmatr)) {
  Cons <- c(Cons, names(which.max(Prof[,i])))
}

paste(Cons, collapse = "")
Prof

A <- c(paste("A:"), paste(Prof["A",]))
C <- c(paste("C:"), paste(Prof["C",]))
G <- c(paste("G:"), paste(Prof["G",]))
TT <- c(paste("T:"), paste(Prof["T",]))
write(paste(Cons, collapse = ""), file = "DNAmat.txt", ncolumns = length(A), append = FALSE)
write(paste(A), file = "DNAmat.txt", ncolumns = length(A), append = TRUE)
write(paste(C), file = "DNAmat.txt", ncolumns = length(C), append = TRUE)
write(paste(G), file = "DNAmat.txt", ncolumns = length(G), append = TRUE)
write(paste(TT), file = "DNAmat.txt", ncolumns = length(TT), append = TRUE)

#----------------

#Alternatively
#Read data
dna <- read.table("rosalind_cons.txt",as.is=T)
acgt <- matrix(c("A","C","G","T"))

ix <- which(sapply(dna, nchar) == 14)
dna <- as.matrix(dna[-ix,])

dna <- t(apply(matrix(1:dim(dna)[1]),1,function(x){strsplit(dna[x,], NULL)[[1]]}))


#Make Profile
profile <- apply(acgt,1,function(x){
            apply(ifelse(dna==x,1,0),2,sum)
                        })
profile <- apply(acgt, 1,function(x){
            sapply(ifelse(sapply(dna,match,x),1,0),sum)
                        })

profile<-t(profile)
rownames(profile)<-acgt

#Make Consensus
c<-matrix(apply(profile,2,max))
consensus <- paste(apply(matrix(1:dim(dna)[2]),1,function(x){rownames(profile)[profile[,x]==c[x]][1]}),collapse="")

#Output Results
rownames(profile)<- c("A:","C:","G:","T:")
write.table(consensus,"output.txt",row.names=F,col.names=F,quote = F)
write.table(profile,"output.txt",row.names=T,col.names=F,quote = F,append=T)


#Alternatively 
alphabet = c("A","C","G","T")
mat = read.table("rosalind_cons.txt")

smat <- paste(mat$V1, collapse = "")
smat <- strsplit(smat, ">")
smat <- unlist(smat)
smat <- smat[-1]
smat <- gsub("Rosalind_[0-9]*","",smat)
smat <- cbind(smat)

mat2 <- t(apply(smat, 1, function(x){ strsplit(x, "")[[1]] }))

res=apply(mat2,2,function(x){  a=sum(x=="A"); c=sum(x=="C"); g=sum(x=="G"); t=sum(x=="T"); return(c(a,c,g,t)) })
rownames(res)=c("A:","C:","G:","T:")
consensus = apply(res,2,function(x){  alphabet[which.max(x)] })
consensus = paste(consensus,collapse="")
```

####Mortal Fibonacci rabbits
```{r Mortal Fibonacci rabbits, message=FALSE, warning=FALSE}

rF <- c(1,1) #Vector of rabbits alive each month
rF <- as.bigz(rF)
n <- 86 #Number of i total
m <- 17 #Number of i alive
k = 1
i = 1
for(i in 3:n) {
    if(i > 2 && m >= i) { #3 to 4, two elem (m=4)
    rF[i] = sum(as.bigz(rF[i-1]), as.bigz(rF[i-2]))
  } else if(i == m+1) { #5
    rF[i] = sum(as.bigz(rF[i-1]),as.bigz(rF[i-2]),-1)
  } else if(i >= m+2 && i <m+4) { #6 to 7
    rF[i] = sum(as.bigz(rF[i-1]),as.bigz(rF[i-2]),-1)
    } else rF[i] = sum(as.bigz(rF[i-1]),as.bigz(rF[i-2]),as.bigz(-rF[i-m-1])) 
  }
as.bigz(rF[length(rF)])
#rF[length(rF)]
```

```{r}
#Correctly:
library('gmp')

rabbitpairs <- function(n,m) {

adult   <- as.bigz(vector(length = n))
adult[1:2] <- c(as.bigz("0"),as.bigz("1"))
offsp   <- as.bigz(vector(length = n))
offsp[1:2] <- c(as.bigz("1"),as.bigz("0"))

seq <- numeric(length = n)

for(i in 3:n){

    adult[i] <- adult[i-1] + offsp[i-1]

    if(i > m ){         
        adult[i] <- adult[i] - offsp[i-m]
    }

    offsp[i] <- adult[i-1]
    }

seq <- adult + offsp
seq[n]
}
rabbitpairs(86,17)
```

